### DISCUSSION TITLE
Feature request: set retention for uploaded files

### DISCUSSION THREAD

**Discussion by itaybar** [open-webui/open-webui#5199] [Ideas]
I think it would be great to add retention time for users upload to prevent out of storage in the DATA DIR

**Comment by rhabichl** [comment]
Our company would need this, also for conversational data, because of privacy concerns. For example delete everything that is older then 30d.

**Reply by surak** [reply]
In fact, everything older than a conversation would be awesome. I don't want dangling pdfs of my server ever

**Comment by surak** [comment]
Is there any function to cleanup the DATA_DIR file in case of disk full, for example, other than deleting it and relaunching webui? I’ve been looking at the documentation and source code and couldn’t find a thing. 

Having a maximum age of data would solve this, too. Say files and entries older than X time get wiped out, where X can be an environment variable.

**Reply by Classic298** [reply]
Well you can write a cron job to automatically 1) stop OWUI 2) delete upload folder 3) restart

And what you can also do is connect to the chromadb and sqlite db and search for orphaned files that no longer belong to an existing chat and embeddings no longer belonging to files and delete them all

**Reply by surak** [reply]
That supposes downtime, and it's a hack. I was thinking something more proper, like periodic automatic cleanup by the system itself.

**Reply by Classic298** [reply]
Well yeah but the downtime is absolutely minimal. Just do it at midnight. It's 2minutes of downtime maximum

And maybe it's a hack but if you build it properly it wouldn't be any worse than an official implementation

**Comment by m8dhouse** [comment]
I wrote a cleanup script a while back. This needs to be checkecd as I'm not following the DB changes.
You could use this as a basis: [openwebui-scripts GitHub](https://github.com/m8dhouse/openwebui-scripts)

It checks files, what has been used in chats, etc ... 
Has a test execution flag where it does nothing but gives you the report.

**Reply by Classic298** [reply]
haha, yes I saw your script. I didnt use your script at all because it was too short and probably outdated though.
I wrote my own based on current DB structure. I am not 100% sure whether it works as intended, _**BUT**_ it _**DOES**_ work.
What i mean by that: I _**think**_ I am always deleting all the embeds, not just the orphaned ones... but I am not 100% sure either.
But oh well, it does what I want it to do at it's core, which is deleting uploaded files and embeds and file handles in the chroma db and main database, so eh, it does what I want and thats good enough.
Maybe I can open a PR to improve your scripts a bit 😁

**Reply by surak** [reply]
Could you share your script somehow? It would be most appreciated!

**Reply by Ithanil** [reply]
@Classic298 Yes, it would be lovely and most appreciated if you could share your efforts. We should make something like this public and maintained.

**Reply by mahenning** [reply]
I wrote a cleanup script for unused files for sqlite3/chromadb in https://github.com/open-webui/open-webui/discussions/12280#discussioncomment-12791253, with an option for dry run (actually you have to set explicit delete parameter, otherwise it just lists candidates). It does not have a time limit for old files in it, but the option to clear the upload dir/files db/vector db separately.

**Reply by Ithanil** [reply]
If you use PG/pgvector, you can find my cleanup script here: https://github.com/open-webui/open-webui/discussions/12091#discussioncomment-12792278

Also has `--dry-run`, but at the moment default is *not* `--dry-run`.

**Comment by spammenotinoz** [comment]
This is a really good idea

**Comment by flefevre** [comment]
Equivalent posts in order to help to gather same idea and consolidate them into a common proposal 

Request for Transparency and Management Options for Audio File Retention #6705 

Enhancement: Add configurable retention and storage location for meeting audio notes #14129

**Comment by Classic298** [comment]
Hello everyone.

I worked on a PR related to this discussion.
Testing wanted!

https://github.com/open-webui/open-webui/pull/16520

**Reply by flefevre** [reply]
That's definitively rocks !

**Comment by Classic298** [comment]
https://github.com/Classic298/prune-open-webui
